analytics/tracker.go

```diff
@@ -0,0 +1,135 @@
+	Wait()
+}
+
+type tracker struct {
+	tracker       analytics.Tracker
+	envRepository env.Repository
+}
+
+// NewTracker ...
+func NewTracker() Tracker {
+	if isAnalyticsDisabled() {
+		return NoOpTracker{}
+	}
+	return tracker{
+		tracker:       analytics.NewDefaultTracker(),
+		envRepository: env.NewRepository(),
+	}
+}
+
+// SendWorkflowStarted ...
+func (t tracker) SendWorkflowStarted(properties analytics.Properties, title string) {
+	isCI := t.envRepository.Get(configs.CIModeEnvKey) == "true"
+	isPR := t.envRepository.Get(configs.PRModeEnvKey) == "true"
+	isDebug := t.envRepository.Get(configs.DebugModeEnvKey) == "true"
+	isSecretFiltering := t.envRepository.Get(configs.IsSecretFilteringKey) == "true"
+	isSecretEnvsFiltering := t.envRepository.Get(configs.IsSecretEnvsFilteringKey) == "true"
+	stateProperties := analytics.Properties{
+		"workflow_name":         title,
+		"ci_mode":               isCI,
+		"pr_mode":               isPR,
+		"debug_mode":            isDebug,
+		"secret_filtering":      isSecretFiltering,
+		"secret_envs_filtering": isSecretEnvsFiltering,
+	}
+	stateProperties.AppendIfNotEmpty("build_slug", t.envRepository.Get("BITRISE_BUILD_SLUG"))
+	stateProperties.AppendIfNotEmpty("parent_step_unique_id", t.envRepository.Get("BITRISE_STEP_UNIQUE_ID"))
+	stateProperties.AppendIfNotEmpty("log_level", t.envRepository.Get(configs.LogLevelEnvKey))
+	currentVersionMap, err := version.ToolVersionMap(os.Args[0])
+	if err == nil {
+		bitriseVersion := currentVersionMap["bitrise"]
+		envmanVersion := currentVersionMap["envman"]
+		stepmanVersion := currentVersionMap["stepman"]
+		stateProperties.AppendIfNotEmpty("cli_version", bitriseVersion.String())
+		stateProperties.AppendIfNotEmpty("envman_version", envmanVersion.String())
+		stateProperties.AppendIfNotEmpty("stepman_version", stepmanVersion.String())
+	} else {
+		log.Debugf("Couldn't get tool versions: %s", err.Error())
+	}
+	t.tracker.Enqueue("workflow_started", properties, stateProperties)
+}
+
+// SendWorkflowFinished ...
+func (t tracker) SendWorkflowFinished(properties analytics.Properties, failed bool) {
+	var statusMessage string
+	if failed {
+		statusMessage = "failed"
+	} else {
+		statusMessage = "successful"
+	}
+	t.tracker.Enqueue("workflow_finished", properties, analytics.Properties{"status": statusMessage})
+}
+
+// SendStepStartedEvent ...
+func (t tracker) SendStepStartedEvent(properties analytics.Properties, infoModel stepmanModels.StepInfoModel, inputs map[string]interface{}) {
+	inputBytes, err := json.Marshal(inputs)
+	if err != nil {
+		log.Errorf("Failed to marshal inputs: %s", err)
+	}
+	t.tracker.Enqueue("step_started", properties, prepareStartProperties(infoModel), analytics.Properties{"inputs": string(inputBytes)})
+}
+
+// SendStepFinishedEvent ...
+func (t tracker) SendStepFinishedEvent(properties analytics.Properties, results models.StepRunResultsModel) {
+	switch results.Status {
```

Comment:
what about separating the Enqueue params building phase from the actually Enqueue call?

So that we could easier review the differences based on the Step run status.